src/nnet3bin/nnet3-combine.cc

```diff
@@ -75,6 +121,10 @@ int main(int argc, char *argv[]) {
 
     Nnet nnet;
     ReadKaldiObject(nnet_rxfilename, &nnet);
+    Nnet moving_average_nnet(nnet), best_nnet(nnet);
```

Comment:
You need to be a bit careful about when these nnets are copied and when they have batchnorm and dropout test-mode set.  We don't want the nnet we write to have these test modes altered vs. the models we read in, but on the other hand we do want these test-modes to be set- if specified- when we measure the objective function.  The test-modes will be copied any time you copy (or assign) the neural nets.

It would be easiest (the least hassle) to have the variables deciding whether the test-modes are set, passed into the function that computes the objective function, and if at least one of those variables is set, have that function copy the model and set the test-modes, and then recurse (this will keep the code simple-- you'll see when you implement it).

Reply:
@danpovey I am a bit confused about this: since I cannot modify NnetComputeProb::nnet_ after initializing an object of NnetComputeProb, I have to create a new object within ComputeObjf() to associate it with the copied model(like the code shown below), which seems inefficient as we would call ComputeObjf() many times. Did I miss something?

`double ComputeObjf(bool batchnorm_test_mode, bool dropout_test_mode, const std::vector<NnetExample> &egs, Nnet &moving_average_nnet, NnetComputeProb *prob_computer) { `

     if (batchnorm_test_mode || dropout_test_mode) {
       Nnet nnet_copy(moving_average_nnet);
       set batchnorm/dropout test mode to nnet_copy
       NnetComputeProb prob_computer_new(compute_prob_opts, nnet_copy);
       return ComputeObjf(false, false, egs, nnet_copy, &prob_computer_new);
     } else {
       compute objf 
       return objf;
     }
}

Reply:
I think creating that object many times is OK.
The only thing it's re-doing is graph compilation, which takes much less than a second; and the maximum number of times it will do this is 30 (assuming you do that other thing I was talking about).

Reply:
OK, Thanks